### PR TITLE
Fix CSR Loss and update training script

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

## Pull Request overview
* Patch the CSRLoss in a few locations
* Update the CSRLoss training script
* Add torch.inference_mode to SparseEncoder & SentenceTransformer

## Details
There's a lot of niche details hidden in the changes, that I'll showcase with comments on specific lines below.
This is the "final" model with the training script: https://huggingface.co/tomaarsen/csr-mxbai-embed-large-v1-nq

- Tom Aarsen

### CODE REVIEW

File: examples/sparse_encoder/training/retrievers/train_csr_nq.py [hunk 1]
```diff
@@ -61,13 +64,31 @@ def main():
...
 
-    # 3. Define our training loss.
-    loss = losses.CSRLoss(model=model, loss=losses.SparseMultipleNegativesRankingLoss(model=model))
+    # 3. Define our training loss. We use the Cosine Similarity similarity as the pretrained model was also trained with it.
+    # The scale of 20 is common for MultipleNegativesRankingLoss with cosine similarity.
+    # The lower gamma gives higher weight to the high-sparsity performance, whereas a higher gamma would give more weight
+    # to the low-sparsity performance.
+    loss = losses.CSRLoss(
+        model=model,
+        loss=losses.SparseMultipleNegativesRankingLoss(model=model, scale=20.0, similarity_fct=util.cos_sim),
```
Comment (tomaarsen) [inline examples/sparse_encoder/training/retrievers/train_csr_nq.py:73]: I got much stronger performance with cosine similarity and scale=20 than dot score, perhaps because the base model uses cosine similarity? You can check my HF for other experiments.

File: examples/sparse_encoder/training/retrievers/train_csr_nq.py [hunk 2]
```diff
@@ -61,13 +64,31 @@ def main():
...
+        gamma=0.1,
+    )
 
     # 4. Define evaluator. This is useful to keep track of alongside the evaluation loss.
     evaluators = []
-    for k_dim in [128, 256]:
-        evaluators.append(evaluation.SparseNanoBEIREvaluator(["msmarco", "nfcorpus", "nq"], max_active_dims=k_dim))
+    for k_dim in [4, 8, 16, 32, 64, 128, 256]:
+        queries = dict(enumerate(eval_dataset["query"]))
```
Comment (tomaarsen) [inline examples/sparse_encoder/training/retrievers/train_csr_nq.py:80]: Out of domain performance was rather weak, even after getting pretty useful performance on the eval set of NQ, see https://huggingface.co/tomaarsen/csr-mxbai-embed-large-v1-nq-nanobeir (this is the same model, but eval'd on NanoBEIR instead)

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 3]
```diff
@@ -576,17 +576,18 @@ def encode(
...
-                if save_to_cpu:
-                    embeddings = embeddings.cpu()
 
-                all_embeddings.extend(embeddings)
+            if convert_to_sparse_tensor:
+                embeddings = embeddings.to_sparse()
+            if save_to_cpu:
+                embeddings = embeddings.cpu()
+
+            all_embeddings.extend(embeddings)
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:590]: Due to what seems to be a bug in torch, I get issues with `coalesce` over the sparse embeddings if the `to_sparse` call happens inside of the inference-mode, so the inference-mode doesn't include it. The select_max_active_dims was required to be in the inference-mode, I get an error otherwise.

P.s. inference_mode also implies no_grad.

File: sentence_transformers/sparse_encoder/losses/CSRLoss.py [hunk 4]
```diff
@@ -91,24 +95,26 @@ def compute_loss_from_embeddings(self, outputs: list[dict[str, torch.Tensor]]) -
...
-        total_loss = total_L_k + total_L_4k / 8 + self.beta * total_L_aux
-
-        return total_loss
+        num_columns = len(outputs)
+        if num_columns > 0:
+            total_L_k /= num_columns
+            total_L_4k /= num_columns
+            total_L_aux /= num_columns
+
+        # return the total losses as a dictionary, they'll be summed for a final reconstruction loss
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/losses/CSRLoss.py:112]: I guess we can just return a big dictionary instead of summing it together? I'm also down to keep it as it was to have just 1 logged reconstruction loss, but I think it's interesting to see the components.

Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/losses/CSRLoss.py:112]: It's definitely better like this. It was made before we start to actually handle multiple losses as a dict

File: sentence_transformers/sparse_encoder/models/CSRSparsity.py [hunk 5]
```diff
@@ -179,7 +190,7 @@ def forward(self, features: torch.Tensor, max_active_dims: int | None = None) ->
                 "decoded_embedding_k": recons_k,
                 "decoded_embedding_4k": recons_4k,
                 "decoded_embedding_aux": recons_aux,
-                "decoded_embedding_k_pre_bias": recons_k + self.pre_bias,
+                "decoded_embedding_k_pre_bias": recons_k - self.pre_bias,
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/models/CSRSparsity.py:193]: Good catch !

File: examples/sparse_encoder/training/retrievers/train_csr_nq.py (hunk 1 above)
Comment (arthurbr11) [reply examples/sparse_encoder/training/retrievers/train_csr_nq.py:73]: Probably yes ! Thanks for noticing

File: examples/sparse_encoder/training/retrievers/train_csr_nq.py (hunk 2 above)
Comment (arthurbr11) [reply examples/sparse_encoder/training/retrievers/train_csr_nq.py:80]: I feel like they seems to be weak but not that much for bigger dimensions like 218 or 256 which is still a good efficiency compare to the original dimension

**Review by arthurbr11 — APPROVED** [review]
Thanks for the PR ! They were indeed some bug in the implementation !
